boxoffice/models/line_item.py

```diff
@@ -222,49 +223,72 @@ def counts_per_date_per_item(item_collection, user_tz):
+    end_at = midnight_to_utc(sales_datetime + datetime.timedelta(days=1), timezone=timezone)
+    sales_on_date = db.session.query('sum').from_statement('''SELECT SUM(final_amount) FROM line_item
+        WHERE status=:status AND ordered_at >= :start_at AND ordered_at < :end_at
+        AND line_item.item_id IN :item_ids
+        ''').params(status=LINE_ITEM_STATUS.CONFIRMED, start_at=start_at, end_at=end_at, item_ids=tuple(item_ids)).scalar()
+    return sales_on_date if sales_on_date else Decimal(0)
 
-    for sales_date in dates:
-        sales_on_date = db.session.query('sum').from_statement('''SELECT SUM(final_amount) FROM line_item
-            WHERE status=:status AND DATE_TRUNC('day', line_item.ordered_at AT TIME ZONE 'UTC' AT TIME ZONE :timezone)::date = :sales_date
-            AND line_item.item_id IN :item_ids
-            ''').params(timezone=user_tz, status=LINE_ITEM_STATUS.CONFIRMED, sales_date=sales_date, item_ids=tuple(item_ids)).first()
-        date_sales[sales_date] = sales_on_date[0] if sales_on_date[0] else Decimal(0)
-    return date_sales
 
-
-def sales_delta(user_tz, item_ids):
+def calculate_weekly_sales(item_collection_ids, user_tz, year):
     """
-    Calculates the percentage difference in sales between today and yesterday
+    Calculates sales per week for items in the given set of item_collection_ids in a given year,
+    in the user's timezone.
     """
-    today = datetime.datetime.now().strftime("%Y-%m-%d")
-    yesterday = (datetime.datetime.now() - datetime.timedelta(days=1)).strftime("%Y-%m-%d")
-    sales = sales_by_date([today, yesterday], user_tz, item_ids)
-    if not sales or not sales[yesterday]:
+    ordered_week_sales = OrderedDict()
+    for year_week in Week.weeks_of_year(year):
+        ordered_week_sales[year_week.week] = 0
+    start_at = isoweek_datetime(year, 1)
+    end_at = isoweek_datetime(year + 1, 1)
+
+    week_sales = db.session.query('sales_week', 'sum').from_statement(db.text('''SELECT EXTRACT(WEEK FROM ordered_at)
+        AS sales_week, SUM(final_amount) AS sum
+        FROM line_item INNER JOIN item on line_item.item_id = item.id
+        WHERE status=:status AND item_collection_id IN :item_collection_ids
+        AND ordered_at >= :start_at AND ordered_at < :end_at
+        GROUP BY sales_week ORDER BY sales_week;
+        ''')).params(timezone=user_tz, status=LINE_ITEM_STATUS.CONFIRMED,
+        start_at=start_at, end_at=end_at, item_collection_ids=tuple(item_collection_ids)).all()
+
+    for week_sale in week_sales:
+        ordered_week_sales[int(week_sale.sales_week)] = week_sale.sum
+
+    return ordered_week_sales
+
+
+def sales_delta(user_tz, item_ids):
+    """Calculates the percentage difference in sales between today and yesterday"""
+    today = midnight_to_utc(datetime.datetime.utcnow(), timezone=user_tz)
+    yesterday = midnight_to_utc(datetime.datetime.utcnow() - datetime.timedelta(days=1), timezone=user_tz)
```

Comment:
Rather than call `utcnow()` twice and get slightly different values, call it once and assign to a variable, and use the variable both times. Convention is to name this variable `now`. Doing this saves you from a suble timing attack when you're milliseconds from midnight and the two calls end up on different days.